### PR TITLE
feat: Workflow — Send Email to Specific Email

### LINKED ISSUE #14466: [CAL-3372] Workflow — Send Email to Specific Email
@lorenzo and I are working on getting calls booked for @peer + @baileypumfleet — it would be great to be notified when a call actually gets booked. 

I think this can be accomplished via the "Workflows" by adding "Send email to specific email" similar to how we do it for "Send text to specific number" 

I guess for now, I could add Lorenzo and my number to get text notifications when calls are booked as a work around

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3372](https://linear.app/calcom/issue/CAL-3372/workflow-send-email-to-specific-email)</sub>

### PR DESCRIPTION
**Pull request by Amit91848**
## What does this PR do?

Fixes #14466 

I don't have sendgrid account cannot show the email, instead showing the logs.

https://github.com/calcom/cal.com/assets/74371312/26a9a406-4931-4f86-b266-da45a5f59fd3

### **Pending**
- [x] ~~Table for Verified Emails like there is of VerifiedNumber~~ 

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

### CODE REVIEW

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
Great work so far, I left you some comments that need to be addressed first 🙏

And I think the implemenation for EMAIL_ADRESSE is missing in [activateEventType.handler.ts](https://github.com/calcom/cal.com/blob/main/packages/trpc/server/routers/viewer/workflows/activateEventType.handler.ts#L190)

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 1]
```diff
@@ -39,9 +39,49 @@ type UpdateOptions = {
...
+      OR: [{ userId }, { teamId }],
+    },
+  });
+
+  if (!verifiedEmail) {
+    throw new TRPCError({ code: "NOT_FOUND", message: "Email not verified" });
+  }
+
+  if (teamId) {
+    if (!verifiedEmail.teamId) {
```
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: If is team workflow but email was only verified for user, then add it to team as well.

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 2]
```diff
@@ -39,9 +39,49 @@ type UpdateOptions = {
...
+      await prisma.verifiedEmail.update({
+        where: {
+          id: verifiedEmail.id,
+        },
+        data: {
+          teamId,
+        },
+      });
+    } else if (verifiedEmail.teamId !== teamId) {
+      await prisma.verifiedEmail.create({
```
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:70]: Suppose user has multiple teams and email was verified for User workflow and then was added to Team A, but now user also used same email for Team B, then create a new entry for Team B

**Comment by Amit91848** [comment]
Made changes @CarinaWolli

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts (hunk 1 above)
Comment (CarinaWolli) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: `verifyEmailSender` is only called when reminders are being scheduled which is not always the case

Comment (CarinaWolli) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: `verifyEmailSender` is only called when new reminders are scheduled which is not always the case

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 3]
```diff
@@ -425,17 +466,16 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
         throw new TRPCError({ code: "UNAUTHORIZED", message: "Not available on free plan" });
       }
       const requiresSender =
-        newStep.action === WorkflowActions.SMS_NUMBER || newStep.action === WorkflowActions.WHATSAPP_NUMBER;
+        newStep.action === WorkflowActions.SMS_NUMBER ||
+        newStep.action === WorkflowActions.WHATSAPP_NUMBER ||
+        WorkflowActions.EMAIL_ADDRESS;
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:471]:
```suggestion
        newStep.action === WorkflowActions.EMAIL_ADDRESS;
```

**Review by CarinaWolli — CHANGES_REQUESTED** [review]
- Right now not verified emails can be saved to the workflow, we should add an error message like for SMS to specific number: 
Screenshot 2024-05-09 at 12 50 38 PM

- Can we change the title of the email form 'Cal.com: Verify your account' to 'Cal.com: Verify your email'

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts (hunk 1 above)
Comment (Amit91848) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: Not sure wdym by that. I called it everywhere we were using sendTo for EMAIL_ADDRESS action.

Comment (CarinaWolli) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: We don't call `verifyEmailSender` when no reminders are scheduled. So for example you can do that: 

1. Create workflow with 'When event is booked' trigger 
2. Choose 'send Email to specific email address' 
3. Type a not verified email address 
4. Type random numbers as the verification code
5. Click save 

This now successfully saves the workflow without `verifyEmailSender` ever being called.

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 4]
```diff
@@ -39,6 +39,45 @@ type UpdateOptions = {
...
+) => {
+  const verifiedEmail = await prisma.verifiedEmail.findFirst({
+    where: {
+      email,
+      OR: [{ userId }, { teamId }],
+    },
+  });
+
+  if (!verifiedEmail) {
+    throw new TRPCError({ code: "NOT_FOUND", message: "Email not verified" });
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:56]: Let's show a toast with this error message when this error thrown

**Comment by Amit91848** [comment]
> > Also cannot open workflow page, keep getting this error.
> 
> For the` getStaticProps()` error, can you try to set the env variable `APP_ROUTER_WORKFLOWS_ENABLED` to 0

Was already 0, tried with 1 also, still getting the same error

**Comment by CarinaWolli** [comment]
> > > Also cannot open workflow page, keep getting this error.
> > 
> > 
> > For the` getStaticProps()` error, can you try to set the env variable `APP_ROUTER_WORKFLOWS_ENABLED` to 0
> 
> Was already 0, tried with 1 also, still getting the same error

did you restart the `yarn dev` server after changing the env. variable? For me it works, I am getting the same error if I set the env variable to 1

**Comment by Amit91848** [comment]
> did you restart the yarn dev server after changing the env. variable? For me it works, I am getting the same error if I set the env variable to 1

Yup still getting the same error


https://github.com/calcom/cal.com/assets/74371312/9ab83cfd-7ad2-4d5c-8dff-ec8d57172c58



Works if i comment this tho

https://github.com/calcom/cal.com/blob/56f31eee7ab517c844b6ad5928dfebb7d438626e/apps/web/middleware.ts#L169

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts (hunk 1 above)
Comment (Amit91848) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:60]: Yeah should have been checking before submitting form fixed this.

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts (hunk 4 above)
Comment (Amit91848) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:56]: https://github.com/calcom/cal.com/blob/56f31eee7ab517c844b6ad5928dfebb7d438626e/packages/features/ee/workflows/pages/workflow.tsx#L198-L204

Should be handled here

**Comment by Amit91848** [comment]
Made changes @CarinaWolli

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 5]
```diff
@@ -309,8 +348,10 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
                 case WorkflowActions.EMAIL_ATTENDEE:
                   sendTo = bookingInfo.attendees.map((attendee) => attendee.email);
                   break;
-                /*case WorkflowActions.EMAIL_ADDRESS:
-                      sendTo = step.sendTo || "";*/
+                case WorkflowActions.EMAIL_ADDRESS:
+                  await verifyEmailSender(step.sendTo || "", user.id, userWorkflow.teamId, ctx.prisma);
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:352]: We are still having the issue that the email doesn't show as verified for a team if a user created it with a verified user email. This `verifyEmailSender` is never called when no reminders for already existing bookings need to be scheduled.


We should call `verifyEmailSender` in a different place. Let's call it when a new step was added and when a step was edited

**Review by CarinaWolli — CHANGES_REQUESTED** [review]

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts (hunk 5 above)
Comment (Amit91848) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts:352]: Missed it here, thats why it was never being called when there are no reminders.

https://github.com/calcom/cal.com/blob/f4927a39df89af51d7d108cce1c2e6d1e867efba/packages/trpc/server/routers/viewer/workflows/update.handler.ts#L463-L475

Also throwing error leads to unhandled rejections. Although further steps aren't processed, but the response is shown as 200

Screenshot 2024-05-17 at 10 37 51 PM

File: packages/trpc/server/routers/viewer/workflows/update.handler.ts [hunk 6]
```diff
@@ -677,8 +722,10 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
                 case WorkflowActions.EMAIL_ATTENDEE:
                   sendTo = bookingInfo.attendees.map((attendee) => attendee.email);
                   break;
-                /*case WorkflowActions.EMAIL_ADDRESS:
-                      sendTo = step.sendTo || "";*/
+                case WorkflowActions.EMAIL_ADDRESS:
+                  await verifyEmailSender(newStep.sendTo || "", user.id, userWorkflow.teamId, ctx.prisma);
```
Comment (CarinaWolli) [inline packages/trpc/server/routers/viewer/workflows/update.handler.ts:726]: for added steps we only call it when we have bookings that need reminders scheduled. We should do it the same as we do for the edited steps

Comment (Amit91848) [reply packages/trpc/server/routers/viewer/workflows/update.handler.ts]: Yeah removed it from here and added when creating new step

**Review by CarinaWolli — APPROVED** [review]
Great work, all issues are resolved and it is working as expected 🙏

**Pull request merged**